Mgmt queries to struct

I want to convert management query results to struct.

simplified example:
```

type Result struct {
	Command string `kusto:"Command"`
}
```

code:

```
iter, err := client.Mgmt(ctx, databaseName, kusto.NewStmtUnsafe(".show tables | project Command=\"some command\""))
if err != nil {
panic(err)
}

defer iter.Stop()

var recs []Result
err = iter.Do(
func(row *table.Row) error {
	rec := Result{}
	if err := row.ToStruct(&rec); err != nil {
		return err
	}
	recs = append(recs, rec)
	fmt.Println(recs)
	fmt.Println("recs length is", len(recs))
	return nil
},
)
if err != nil {
  fmt.Println("ERROR", err.Error())
  panic(err)
}

```

By running the KQL:

```
.show tables | project Command="some command"
```

Let's say I have 3 tables, so I get 3 lines of "some command".

Code output:

```
[{some command}]
recs length is 1
[{some command} {some command}]
recs length is 2
[{some command} {some command} {some command}]
recs length is 3
[{some command} {some command} {some command} {{"Visualization":null,"Title":null,"XColumn":null,"Series":null,"YColumns":null,"AnomalyColumns":null,"XTitle":null,"YTitle":null,"XAxis":null,"YAxis":null,"Legend":null,"YSplit":null,"Accumulate":false,"IsQuerySorted":false,"Kind":null,"Ymin":"NaN","Ymax":"NaN"}}]
recs length is 4
ERROR Op(OpMgmt): Kind(KClientArgs): row does not have the correct number of values(10) for the number of columns(1)
panic: Op(OpMgmt): Kind(KClientArgs): row does not have the correct number of values(10) for the number of columns(1)
```

It seems like it tries to convert all received rows including the metadata, statistics etc and tries to convert them to the desired struct, which fails.

--- a/kusto/internal/frames/frames.go
+++ b/kusto/internal/frames/frames.go
@@ -56,6 +56,7 @@ const (
 	QueryCompletionInformation TableKind = "QueryCompletionInformation"
 	QueryTraceLog              TableKind = "QueryTraceLog"
 	QueryPerfLog               TableKind = "QueryPerfLog"
+	QueryResult                TableKind = "QueryResult"
 	TableOfContents            TableKind = "TableOfContents"
 	QueryPlan                  TableKind = "QueryPlan"
 	ExtendedProperties         TableKind = "@ExtendedProperties"

--- a/kusto/statemachine.go
+++ b/kusto/statemachine.go
@@ -7,9 +7,8 @@ import (
 	"context"
 	"sync"
 
-	"github.com/Azure/azure-kusto-go/kusto/data/table"
-
 	"github.com/Azure/azure-kusto-go/kusto/data/errors"
+	"github.com/Azure/azure-kusto-go/kusto/data/table"
 	"github.com/Azure/azure-kusto-go/kusto/internal/frames"
 	v1 "github.com/Azure/azure-kusto-go/kusto/internal/frames/v1"
 	v2 "github.com/Azure/azure-kusto-go/kusto/internal/frames/v2"
@@ -316,11 +315,20 @@ type v1SM struct {
 	columnSetOnce sync.Once
 	ctx           context.Context
 
-	currentFrame frames.Frame
+	currentTable v1.DataTable
+	tables       []v1.DataTable
 
 	receivedDT bool
 
 	wg *sync.WaitGroup
+}
+
+type TableOfContents struct {
+	Ordinal    int64
+	Kind       string
+	Name       string
+	Id         string
+	PrettyName string
 }
 
 func (p *v1SM) start() (stateFn, error) {
@@ -337,26 +345,21 @@ func (p *v1SM) nextFrame() (stateFn, error) {
 		return nil, p.ctx.Err()
 	case fr, ok := <-p.in:
 		if !ok {
-			if !p.receivedDT {
-				return nil, errors.ES(p.op, errors.KInternal, "received a table stream that did not finish before our input channel, this is usally a return size or time limit")
+			if len(p.tables) == 0 {
+				return p.done, nil
 			}
-			p.wg.Wait()
-			return nil, nil
+			if len(p.tables) <= 2 {
+				p.currentTable = p.tables[0]
+			} else {
+				p.currentTable = p.tables[len(p.tables)-1]
+				return p.tableOfContents, nil
+			}
+			return p.dataTable, nil
 		}
-		p.currentFrame = fr
 		switch tbl := fr.(type) {
 		case v1.DataTable:
-			var err error
-			p.columnSetOnce.Do(func() {
-				var cols table.Columns
-				cols, err = tbl.DataTypes.ToColumns()
-				if err != nil {
-					return
-				}
-				p.wg.Add(1)
-				p.iter.inColumns <- send{inColumns: cols, wg: p.wg}
-			})
-			return p.dataTable, err
+			p.tables = append(p.tables, tbl)
+			return p.nextFrame, nil
 		case frames.Error:
 			return nil, tbl
 		default:
@@ -364,17 +367,62 @@ func (p *v1SM) nextFrame() (stateFn, error) {
 		}
 	}
 }
+func (p *v1SM) done() (stateFn, error) {
+	if !p.receivedDT {
+		return nil, errors.ES(p.op, errors.KInternal, "received a table stream that did not finish before our input channel, this is usally a return size or time limit")
+	}
+	p.wg.Wait()
+
+	return nil, nil
+}
+
+func (p *v1SM) tableOfContents() (stateFn, error) {
+	tableOfContents := p.currentTable
+	columns, err := tableOfContents.DataTypes.ToColumns()
+	if err != nil {
+		return nil, err
+	}
+
+	current := TableOfContents{}
+	for _, kustoRow := range tableOfContents.KustoRows {
+		row := table.Row{ColumnTypes: columns, Values: kustoRow, Op: p.op}
+		err := row.ToStruct(&current)
+		if err != nil {
+			return nil, err
+		}
+
+		kind := frames.TableKind(current.Kind)
+		if kind == frames.QueryResult {
+			p.currentTable = p.tables[current.Ordinal]
+			if _, err := p.dataTable(); err != nil {
+				return nil, err
+			}
+		}
+	}
+	return p.done, nil
+}
 
 func (p *v1SM) dataTable() (stateFn, error) {
-	table := p.currentFrame.(v1.DataTable)
+	var err error
+	currentTable := p.currentTable
+
+	p.columnSetOnce.Do(func() {
+		var cols table.Columns
+		cols, err = currentTable.DataTypes.ToColumns()
+		if err != nil {
+			return
+		}
+		p.wg.Add(1)
+		p.iter.inColumns <- send{inColumns: cols, wg: p.wg}
+	})
 
 	p.wg.Add(1)
 	select {
 	case <-p.ctx.Done():
 		return nil, p.ctx.Err()
-	case p.iter.inRows <- send{inRows: table.KustoRows, wg: p.wg}:
+	case p.iter.inRows <- send{inRows: currentTable.KustoRows, wg: p.wg}:
 		p.receivedDT = true
 	}
 
-	return p.nextFrame, nil
+	return p.done, nil
 }

--- a/kusto/statemachine_test.go
+++ b/kusto/statemachine_test.go
@@ -776,6 +776,13 @@ func TestV1SM(t *testing.T) {
 					},
 					KustoRows: []value.Values{
 						{
+							value.Long{Value: 0, Valid: true},
+							value.String{Value: "QueryResult", Valid: true},
+							value.String{Value: "PrimaryResult", Valid: true},
+							value.String{Value: "07dd9603-3e06-4c62-986b-dfc3d586b05a", Valid: true},
+							value.String{Value: "", Valid: true},
+						},
+						{
 							value.Long{Value: 1, Valid: true},
 							value.String{Value: "QueryProperties", Valid: true},
 							value.String{Value: "@ExtendedProperties", Valid: true},
@@ -784,13 +791,6 @@ func TestV1SM(t *testing.T) {
 						},
 						{
 							value.Long{Value: 2, Valid: true},
-							value.String{Value: "QueryResult", Valid: true},
-							value.String{Value: "PrimaryResult", Valid: true},
-							value.String{Value: "07dd9603-3e06-4c62-986b-dfc3d586b05a", Valid: true},
-							value.String{Value: "", Valid: true},
-						},
-						{
-							value.Long{Value: 0, Valid: true},
 							value.String{Value: "QueryResult", Valid: true},
 							value.String{Value: "PrimaryResult", Valid: true},
 							value.String{Value: "07dd9603-3e06-4c62-986b-dfc3d586b05a", Valid: true},
@@ -808,8 +808,8 @@ func TestV1SM(t *testing.T) {
 					},
 					Values: value.Values{
 						value.DateTime{Value: nowish, Valid: true},
-						value.String{Value: "DD", Valid: true},
-						value.Long{Value: 101, Valid: true},
+						value.String{Value: "Doak", Valid: true},
+						value.Long{Value: 10, Valid: true},
 					},
 					Op: errors.OpQuery,
 				},
@@ -821,8 +821,8 @@ func TestV1SM(t *testing.T) {
 					},
 					Values: value.Values{
 						value.DateTime{Value: nowish, Valid: true},
-						value.String{Value: "Doak", Valid: true},
-						value.Long{Value: 10, Valid: true},
+						value.String{Value: "DD", Valid: true},
+						value.Long{Value: 101, Valid: true},
 					},
 					Op: errors.OpQuery,
 				},

--- a/kusto/statemachine_test.go
+++ b/kusto/statemachine_test.go
@@ -575,7 +575,7 @@ func TestV1SM(t *testing.T) {
 			err:    true,
 		},
 		{
-			desc: "Expected Result",
+			desc: "Single Table",
 			ctx:  context.Background(),
 			stream: []frames.Frame{
 				v1.DataTable{
@@ -589,20 +589,6 @@ func TestV1SM(t *testing.T) {
 							value.DateTime{Value: nowish, Valid: true},
 							value.String{Value: "Doak", Valid: true},
 							value.Long{Value: 10, Valid: true},
-						},
-					},
-				},
-				v1.DataTable{
-					DataTypes: v1.DataTypes{
-						{ColumnName: "Timestamp", ColumnType: "datetime"},
-						{ColumnName: "Name", ColumnType: "string"},
-						{ColumnName: "ID", ColumnType: "long"},
-					},
-					KustoRows: []value.Values{
-						{
-							value.DateTime{Value: nowish, Valid: true},
-							value.String{Value: "Dubovski", Valid: true},
-							value.Long{Value: 0, Valid: false},
 						},
 					},
 				},
@@ -621,6 +607,38 @@ func TestV1SM(t *testing.T) {
 					},
 					Op: errors.OpQuery,
 				},
+			},
+		},
+		{
+			desc: "Primary And QueryProperties",
+			ctx:  context.Background(),
+			stream: []frames.Frame{
+				v1.DataTable{
+					DataTypes: v1.DataTypes{
+						{ColumnName: "Timestamp", ColumnType: "datetime"},
+						{ColumnName: "Name", ColumnType: "string"},
+						{ColumnName: "ID", ColumnType: "long"},
+					},
+					KustoRows: []value.Values{
+						{
+							value.DateTime{Value: nowish, Valid: true},
+							value.String{Value: "Doak", Valid: true},
+							value.Long{Value: 10, Valid: true},
+						},
+					},
+				},
+				v1.DataTable{
+					DataTypes: v1.DataTypes{
+						{ColumnName: "Value", ColumnType: "string"},
+					},
+					KustoRows: []value.Values{
+						{
+							value.String{Value: "{\"Visualization\":null,\"Title\":null,\"XColumn\":null,\"Series\":null,\"YColumns\":null,\"AnomalyColumns\":null,\"XTitle\":null,\"YTitle\":null,\"XAxis\":null,\"YAxis\":null,\"Legend\":null,\"YSplit\":null,\"Accumulate\":false,\"IsQuerySorted\":false,\"Kind\":null,\"Ymin\":\"NaN\",\"Ymax\":\"NaN\"}", Valid: true},
+						},
+					},
+				},
+			},
+			want: table.Rows{
 				&table.Row{
 					ColumnTypes: table.Columns{
 						{Name: "Timestamp", Type: "datetime"},
@@ -629,8 +647,182 @@ func TestV1SM(t *testing.T) {
 					},
 					Values: value.Values{
 						value.DateTime{Value: nowish, Valid: true},
-						value.String{Value: "Dubovski", Valid: true},
-						value.Long{Value: 0, Valid: false},
+						value.String{Value: "Doak", Valid: true},
+						value.Long{Value: 10, Valid: true},
+					},
+					Op: errors.OpQuery,
+				},
+			},
+		},
+		{
+			desc: "Primary With TableOfContents",
+			ctx:  context.Background(),
+			stream: []frames.Frame{
+				v1.DataTable{
+					DataTypes: v1.DataTypes{
+						{ColumnName: "Timestamp", ColumnType: "datetime"},
+						{ColumnName: "Name", ColumnType: "string"},
+						{ColumnName: "ID", ColumnType: "long"},
+					},
+					KustoRows: []value.Values{
+						{
+							value.DateTime{Value: nowish, Valid: true},
+							value.String{Value: "Doak", Valid: true},
+							value.Long{Value: 10, Valid: true},
+						},
+					},
+				},
+				v1.DataTable{
+					DataTypes: v1.DataTypes{
+						{ColumnName: "Value", ColumnType: "string"},
+					},
+					KustoRows: []value.Values{
+						{
+							value.String{Value: "{\"Visualization\":null,\"Title\":null,\"XColumn\":null,\"Series\":null,\"YColumns\":null,\"AnomalyColumns\":null,\"XTitle\":null,\"YTitle\":null,\"XAxis\":null,\"YAxis\":null,\"Legend\":null,\"YSplit\":null,\"Accumulate\":false,\"IsQuerySorted\":false,\"Kind\":null,\"Ymin\":\"NaN\",\"Ymax\":\"NaN\"}", Valid: true},
+						},
+					},
+				},
+				v1.DataTable{
+					DataTypes: v1.DataTypes{
+						{ColumnName: "Ordinal", ColumnType: "long"},
+						{ColumnName: "Kind", ColumnType: "string"},
+						{ColumnName: "Name", ColumnType: "string"},
+						{ColumnName: "Id", ColumnType: "string"},
+						{ColumnName: "PrettyName", ColumnType: "string"},
+					},
+					KustoRows: []value.Values{
+						{
+							value.Long{Value: 0, Valid: true},
+							value.String{Value: "QueryResult", Valid: true},
+							value.String{Value: "PrimaryResult", Valid: true},
+							value.String{Value: "07dd9603-3e06-4c62-986b-dfc3d586b05a", Valid: true},
+							value.String{Value: "", Valid: true},
+						},
+						{
+							value.Long{Value: 1, Valid: true},
+							value.String{Value: "QueryProperties", Valid: true},
+							value.String{Value: "@ExtendedProperties", Valid: true},
+							value.String{Value: "309c015e-5693-4b66-92e7-4a4f98c3155b", Valid: true},
+							value.String{Value: "", Valid: true},
+						},
+					},
+				},
+			},
+			want: table.Rows{
+				&table.Row{
+					ColumnTypes: table.Columns{
+						{Name: "Timestamp", Type: "datetime"},
+						{Name: "Name", Type: "string"},
+						{Name: "ID", Type: "long"},
+					},
+					Values: value.Values{
+						value.DateTime{Value: nowish, Valid: true},
+						value.String{Value: "Doak", Valid: true},
+						value.Long{Value: 10, Valid: true},
+					},
+					Op: errors.OpQuery,
+				},
+			},
+		},
+		{
+			desc: "Multiple Primaries",
+			ctx:  context.Background(),
+			stream: []frames.Frame{
+				v1.DataTable{
+					DataTypes: v1.DataTypes{
+						{ColumnName: "Timestamp", ColumnType: "datetime"},
+						{ColumnName: "Name", ColumnType: "string"},
+						{ColumnName: "ID", ColumnType: "long"},
+					},
+					KustoRows: []value.Values{
+						{
+							value.DateTime{Value: nowish, Valid: true},
+							value.String{Value: "Doak", Valid: true},
+							value.Long{Value: 10, Valid: true},
+						},
+					},
+				},
+				v1.DataTable{
+					DataTypes: v1.DataTypes{
+						{ColumnName: "Value", ColumnType: "string"},
+					},
+					KustoRows: []value.Values{
+						{
+							value.String{Value: "{\"Visualization\":null,\"Title\":null,\"XColumn\":null,\"Series\":null,\"YColumns\":null,\"AnomalyColumns\":null,\"XTitle\":null,\"YTitle\":null,\"XAxis\":null,\"YAxis\":null,\"Legend\":null,\"YSplit\":null,\"Accumulate\":false,\"IsQuerySorted\":false,\"Kind\":null,\"Ymin\":\"NaN\",\"Ymax\":\"NaN\"}", Valid: true},
+						},
+					},
+				},
+				v1.DataTable{
+					DataTypes: v1.DataTypes{
+						{ColumnName: "Timestamp", ColumnType: "datetime"},
+						{ColumnName: "Name", ColumnType: "string"},
+						{ColumnName: "ID", ColumnType: "long"},
+					},
+					KustoRows: []value.Values{
+						{
+							value.DateTime{Value: nowish, Valid: true},
+							value.String{Value: "DD", Valid: true},
+							value.Long{Value: 101, Valid: true},
+						},
+					},
+				},
+				v1.DataTable{
+					DataTypes: v1.DataTypes{
+						{ColumnName: "Ordinal", ColumnType: "long"},
+						{ColumnName: "Kind", ColumnType: "string"},
+						{ColumnName: "Name", ColumnType: "string"},
+						{ColumnName: "Id", ColumnType: "string"},
+						{ColumnName: "PrettyName", ColumnType: "string"},
+					},
+					KustoRows: []value.Values{
+						{
+							value.Long{Value: 1, Valid: true},
+							value.String{Value: "QueryProperties", Valid: true},
+							value.String{Value: "@ExtendedProperties", Valid: true},
+							value.String{Value: "309c015e-5693-4b66-92e7-4a4f98c3155b", Valid: true},
+							value.String{Value: "", Valid: true},
+						},
+						{
+							value.Long{Value: 2, Valid: true},
+							value.String{Value: "QueryResult", Valid: true},
+							value.String{Value: "PrimaryResult", Valid: true},
+							value.String{Value: "07dd9603-3e06-4c62-986b-dfc3d586b05a", Valid: true},
+							value.String{Value: "", Valid: true},
+						},
+						{
+							value.Long{Value: 0, Valid: true},
+							value.String{Value: "QueryResult", Valid: true},
+							value.String{Value: "PrimaryResult", Valid: true},
+							value.String{Value: "07dd9603-3e06-4c62-986b-dfc3d586b05a", Valid: true},
+							value.String{Value: "", Valid: true},
+						},
+					},
+				},
+			},
+			want: table.Rows{
+				&table.Row{
+					ColumnTypes: table.Columns{
+						{Name: "Timestamp", Type: "datetime"},
+						{Name: "Name", Type: "string"},
+						{Name: "ID", Type: "long"},
+					},
+					Values: value.Values{
+						value.DateTime{Value: nowish, Valid: true},
+						value.String{Value: "DD", Valid: true},
+						value.Long{Value: 101, Valid: true},
+					},
+					Op: errors.OpQuery,
+				},
+				&table.Row{
+					ColumnTypes: table.Columns{
+						{Name: "Timestamp", Type: "datetime"},
+						{Name: "Name", Type: "string"},
+						{Name: "ID", Type: "long"},
+					},
+					Values: value.Values{
+						value.DateTime{Value: nowish, Valid: true},
+						value.String{Value: "Doak", Valid: true},
+						value.Long{Value: 10, Valid: true},
 					},
 					Op: errors.OpQuery,
 				},

--- a/kusto/test/etoe/etoe_test.go
+++ b/kusto/test/etoe/etoe_test.go
@@ -38,6 +38,10 @@ type CountResult struct {
 	Count int64
 }
 
+type MgmtProjectionResult struct {
+	A string
+}
+
 type AllDataType struct {
 	Vnum  int32                  `kusto:"vnum"`
 	Vdec  value.Decimal          `kusto:"vdec"`
@@ -159,6 +163,44 @@ func TestQueries(t *testing.T) {
 			gotInit: func() interface{} {
 				return nil
 			},
+		},
+		{
+			desc:  "Mgmt(https://github.com/Azure/azure-kusto-go/issues/55): transformations on mgmt queries",
+			stmt:  kusto.NewStmt(`.show databases | project A="1" | take 1`),
+			mcall: client.Mgmt,
+			doer: func(row *table.Row, update interface{}) error {
+				rec := MgmtProjectionResult{}
+				if err := row.ToStruct(&rec); err != nil {
+					return err
+				}
+				recs := update.(*[]MgmtProjectionResult)
+				*recs = append(*recs, rec)
+				return nil
+			},
+			gotInit: func() interface{} {
+				v := []MgmtProjectionResult{}
+				return &v
+			},
+			want: []MgmtProjectionResult{{A: "1"}},
+		},
+		{
+			desc:  "Mgmt(https://github.com/Azure/azure-kusto-go/issues/55): transformations on mgmt queries - multiple tables",
+			stmt:  kusto.NewStmt(`.show databases | project A="1" | take 1; AllDataTypes | project A="2" | take 1`),
+			mcall: client.Mgmt,
+			doer: func(row *table.Row, update interface{}) error {
+				rec := MgmtProjectionResult{}
+				if err := row.ToStruct(&rec); err != nil {
+					return err
+				}
+				recs := update.(*[]MgmtProjectionResult)
+				*recs = append(*recs, rec)
+				return nil
+			},
+			gotInit: func() interface{} {
+				v := []MgmtProjectionResult{}
+				return &v
+			},
+			want: []MgmtProjectionResult{{A: "1"}, {A: "2"}},
 		},
 		{
 			desc:  "Query: Progressive query: make sure we can convert all data types from a row",

